### PR TITLE
24542 - bad error message for `import *`-ing an `export=`

### LINKED ISSUE #24542: Bad error message for `import *`-ing an `export=`
```ts
// ./b.ts
function foo() {
}
export = foo;
```

```ts
// ./a.ts
import * as b from "./b";
b;
```

```json5
// tsconfig.json
{
    "compilerOptions": {
        "moduleResolution": "node",
        "outDir": "lib"
    },
    "include": [
        "./a.ts"
    ]
}
```

**Expected**:

```
ECMAScript imports can only reference an 'export =' declaration with the 'esModuleInterop' flag enabled and by using default imports.
```

**Actual:**

```
Module '"./b"' resolves to a non-module entity and cannot be imported using this construct.
```

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #24542

### CODE REVIEW

**Comment by mhegazy** [comment]
@DanielRosenwasser can you please review this change.

**Comment by DanielRosenwasser** [comment]
@mhegazy @weswigham what about

```
When writing ECMAScript imports, callable/constructable 'export ='-style modules can only be imported by turning on the 'esModuleInterop' flag and using a default import.
```

**Comment by weswigham** [comment]
@DanielRosenwasser we can equally point people to `import() = require` when not targeting esnext and is _usually_ a less error-prone change than turning on `esModuleInterop`; which is why the original message was "cannot be imported using this construct".

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 1]
```diff
@@ -155,13 +155,13 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
...
                             ~~~~~~~~~~
-!!! error TS2497: Module '"function"' resolves to a non-module entity and cannot be imported using this construct.
+!!! error TS2497: ECMAScript imports can only reference an 'export =' declaration with the 'esModuleInterop' flag enabled and by using default imports.
     import { a as a8 } from "function-module";
     import { a as a9 } from "class";
              ~
 !!! error TS2305: Module '"class"' has no exported member 'a'.
                             ~~~~~~~
-!!! error TS2497: Module '"class"' resolves to a non-module entity and cannot be imported using this construct.
+!!! error TS2497: ECMAScript imports can only reference an 'export =' declaration with the 'esModuleInterop' flag enabled and by using default imports.
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply tests/baselines/reference/es6ExportEqualsInterop.errors.txt]: @DanielRosenwasser This message was provided in the Expected section - #24542. Should I change it?

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser @weswigham should I change the message?

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser @weswigham Should I change the message to 

> When writing ECMAScript imports, callable/constructable 'export ='-style modules can only be imported by turning on the 'esModuleInterop' flag and using a default import.

?

Maybe the message should rely on module type(es2015, cjs), what do you think?

**Comment by DanielRosenwasser** [comment]
Hey @a-tarasyuk - I think you're right, but in those circumstances we need to just recommend people use `allowSyntheticDefaultImports`, not `esModuleInterop`. Does that sound right @weswigham?

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser thanks for your answer, will wait for the answer from @weswigham in order to understand which message should I use for this case as the message from _Expected_ section - #24542 is not right.

**Comment by weswigham** [comment]
> Hey @a-tarasyuk - I think you're right, but in those circumstances we need to just recommend people use allowSyntheticDefaultImports, not esModuleInterop

If module < es2015, esModuleInterop, otherwise alliwSyntheticDefaultImports.

**Comment by mhegazy** [comment]
@DanielRosenwasser can you take a look.

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser could you review the latest changes?

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser could you review the latest changes? Thanks in advance.

**Comment by mhegazy** [comment]
@DanielRosenwasser mentioned he had some comments offline. @DanielRosenwasser can u please comment

**Comment by a-tarasyuk** [comment]
@mhegazy thank you very much., will wait for feedback from @DanielRosenwasser

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser do you have any comments/suggestions?

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser ?

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser ?

**Comment by a-tarasyuk** [comment]
PR was closed, because seems these changes are useless.

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser can we move forward with this "long-live" PR? ;)

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser are there additional comments?

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 2]
```diff
@@ -108,19 +108,19 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
     // namespace import
     import * as y1 from "interface";
                         ~~~~~~~~~~~
-!!! error TS2497: Module '"interface"' resolves to a non-module entity and cannot be imported using this construct.
+!!! error TS2497: When writing ECMAScript imports, callable 'export ='-style modules can only be imported by turning on the 'esModuleInterop' flag and using a default import.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]: I think this error is now somewhat misleading. What about

```
This module can only be imported with ECMAScript imports by turning on the `esModuleInterop` flag and using a default import.
```

Comment (a-tarasyuk) [reply tests/baselines/reference/es6ExportEqualsInterop.errors.txt]: @DanielRosenwasser changed :+1:

Comment (a-tarasyuk) [reply tests/baselines/reference/es6ExportEqualsInterop.errors.txt]: @DanielRosenwasser changed. could you review?

Comment (a-tarasyuk) [reply tests/baselines/reference/es6ExportEqualsInterop.errors.txt]: @DanielRosenwasser changed. could you review?

File: tests/baselines/reference/es6ExportEqualsInterop.errors.txt [hunk 3]
```diff
@@ -190,13 +190,13 @@ tests/cases/compiler/main.ts(106,15): error TS2498: Module '"class-module"' uses
              ~
 !!! error TS2305: Module '"function"' has no exported member 'a'.
                             ~~~~~~~~~~
-!!! error TS2497: Module '"function"' resolves to a non-module entity and cannot be imported using this construct.
+!!! error TS2497: This module can only be imported with ECMAScript imports by turning on the 'esModuleInterop' flag and using a default import.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/es6ExportEqualsInterop.errors.txt]: Damn it, okay, last time and I swear I'l pull it in no matter what.

```
This module can only be referenced with ECMAScript imports/exports by turning on the '{0}' flag and referencing its default export.
```

Comment (a-tarasyuk) [reply tests/baselines/reference/es6ExportEqualsInterop.errors.txt]: @DanielRosenwasser changed. +1:

Comment (a-tarasyuk) [reply tests/baselines/reference/es6ExportEqualsInterop.errors.txt]: @DanielRosenwasser changed. :+1:

Comment (a-tarasyuk) [reply tests/baselines/reference/es6ExportEqualsInterop.errors.txt]: @DanielRosenwasser ok :) :+1:

**Comment by DanielRosenwasser** [comment]
Thank you for the PR and your patience!

**Pull request merged**